test/cc.lean

```diff
@@ -0,0 +1,539 @@
+  cc
+
+example (y1 y2 x2 x3 x4 x5 x6 : α) :
+    (y1 + y2)*x4 = (y2 + y1) → x3*x6 = x5*x5 → x5 = x4 → x6 = x2 →
+      (y2 + y1) = (y1 + y2)*(x6*x3) := by
+  cc
+
+example (y1 y2 y3 x2 x3 x4 x5 x6 : α) :
+    (y1 + y2)*x4 = (y3 + y1) → x3*x6 = x5*x5 → x5 = x4 → x6 = x2 → y2 = y3 →
+      (y2 + y1) = (y1 + y3)*(x6*x3) := by
+  cc
+
+end CCAC5
+
+section CCConstructors
+
+example (a b : Nat) (s t : List Nat) : a :: s = b :: t → a ≠ b → False := by
+  cc
+
+example (a b : Nat) (s t : List Nat) : a :: s = b :: t → t ≠ s → False := by
+  cc
+
+example (a c b : Nat) (s t : List Nat) : a :: s = b :: t → a ≠ c → c = b → False := by
+  cc
+
+example (a c b : Nat) (s t : List Nat) : a :: a :: s = a :: b :: t → a ≠ c → c = b → False := by
+  cc
+
+example (a b : Nat) (s t r : List Nat) : a :: s = r → r = b :: t → a ≠ b → False := by
+  cc
+
+example (a b : Nat) (s t r : List Nat) : a :: s = r → r = b :: t → a = b := by
+  cc
+
+example (a b : Nat) (s t r : List Nat) : List.cons a = List.cons b → a = b := by
+  intro h1
+  /- In the current implementation, `cc` does not "complete" partially applied
+     constructor applications. So, the following one should fail. -/
+  try cc
+  /- Complete it manually. TODO(Leo): we can automate it for inhabited types. -/
+  have h := congr_fun h1 []
+  cc
+
+inductive Foo
+| mk1 : Nat → Nat → Foo
+| mk2 : Nat → Nat → Foo
+
+example (a b : Nat) : Foo.mk1 a = Foo.mk2 b → False := by
+  intro h1
+  /- In the current implementation, `cc` does not "complete" partially applied
+     constructor applications. So, the following one should fail. -/
+  try cc
+  have h := congr_fun h1 0
+  cc
+
+universe u
+inductive Vec (α : Type u) : Nat → Type (max 1 u)
+| nil  : Vec α 0
+| cons : ∀ {n}, α → Vec α n → Vec α (Nat.succ n)
+
+example (α : Type u) (a b c d : α) (n : Nat) (s t : Vec α n) :
+    Vec.cons a s = Vec.cons b t → a ≠ b → False := by
+  cc
+
+example (α : Type u) (a b c d : α) (n : Nat) (s t : Vec α n) :
+    Vec.cons a s = Vec.cons b t → t ≠ s → False := by
+  cc
+
+example (α : Type u) (a b c d : α) (n : Nat) (s t : Vec α n) :
+    Vec.cons a (Vec.cons a s) = Vec.cons a (Vec.cons b t) → b ≠ c → c = a → False := by
+  cc
+
+end CCConstructors
+
+section CCProj
+
+example (a b c d : Nat) (f : Nat → Nat × Nat) : (f d).1 ≠ a → f d = (b, c) → b = a → False := by
+  cc
+
+def ex₂ (a b c d : Nat) (f : Nat → Nat × Nat) : (f d).2 ≠ a → f d = (b, c) → c = a → False := by
+  cc
+
+example (a b c : Nat) (f : Nat → Nat) : (f b, c).1 ≠ f a → f b = f c → a = c → False := by
+  cc
+
+end CCProj
+
+section CCValue
+
+example (a b : Nat) : a = 1 → b = 2 → a = b → False := by
+  cc
+
+example (a b c : Int) : a = 1 → c = -2 → a = b → c = b → False := by
+  cc
+
+example (a b : Char) : a = 'h' → b = 'w' → a = b → False := by
+  cc
+
+example (a b : String) : a = "hello" → b = "world" → a = b → False := by
+  cc
+
+example (a b c : String) : a = c → a = "hello" → c = "world" → c = b → False := by
+  cc
+
+end CCValue
+
```

Comment:
I see you copied most of the tests from the Lean 3 implementation, good idea! I couldn't find the equivalent of the following tests:
```lean
-- tests/run/1442.lean
protected def rel : ℤ × ℤ → ℤ × ℤ → Prop
| ⟨n₁, d₁⟩ ⟨n₂, d₂⟩ := n₁ * d₂ = n₂ * d₁

private def mul' : ℤ × ℤ → ℤ × ℤ → ℤ × ℤ
| ⟨n₁, d₁⟩ ⟨n₂, d₂⟩ := ⟨n₁ * n₂, d₁ * d₂⟩

instance a : is_associative ℤ (*) := ⟨int.mul_assoc⟩

instance c : is_commutative ℤ (*) := ⟨int.mul_comm⟩

example : ∀(a b c d : ℤ × ℤ), rel a c → rel b d → rel (mul' a b) (mul' c d) :=
λ⟨n₁, d₁⟩ ⟨n₂, d₂⟩ ⟨n₃, d₃⟩ ⟨n₄, d₄⟩,
  assume (h₁ : n₁ * d₃ = n₃ * d₁) (h₂ : n₂ * d₄ = n₄ * d₂),
  show (n₁ * n₂) * (d₃ * d₄) = (n₃ * n₄) * (d₁ * d₂),
    by cc
```

```lean
-- tests/run/1608.lean
example {α : Type} {a b : α} (h : ¬ (a = b)) : b ≠ a :=
by cc

example {α : Type} {a b : α} (h : ¬ (a = b)) : ¬ (b = a) :=
by cc
```